### PR TITLE
perf: implement the MultiSet method

### PR DESCRIPTION
**Pull request by j75689**
### Description

Implement the `MultiSet` method and calculate root hash in parallel.

### Rationale

Write to multiple leaf nodes in parallel and calculate Hash upwards at the same time

### Example

```go
testKVData := []Item{
	{1, hasher.Hash([]byte("val1"))},
	{2, hasher.Hash([]byte("val2"))},
	{3, hasher.Hash([]byte("val3"))},
	{4, hasher.Hash([]byte("val4"))},
	{5, hasher.Hash([]byte("val5"))},
	{6, hasher.Hash([]byte("val6"))},
	{7, hasher.Hash([]byte("val7"))},
	{8, hasher.Hash([]byte("val8"))},
	{9, hasher.Hash([]byte("val9"))},
	{10, hasher.Hash([]byte("val10"))},
	{11, hasher.Hash([]byte("val11"))},
	{12, hasher.Hash([]byte("val12"))},
	{13, hasher.Hash([]byte("val13"))},
	{14, hasher.Hash([]byte("val14"))},
	{200, hasher.Hash([]byte("val200"))},
	{20, hasher.Hash([]byte("val20"))},
	{21, hasher.Hash([]byte("val21"))},
	{22, hasher.Hash([]byte("val22"))},
	{23, hasher.Hash([]byte("val23"))},
	{24, hasher.Hash([]byte("val24"))},
	{26, hasher.Hash([]byte("val26"))},
	{37, hasher.Hash([]byte("val37"))},
	{255, hasher.Hash([]byte("val255"))},
	{254, hasher.Hash([]byte("val254"))},
	{253, hasher.Hash([]byte("val253"))},
	{252, hasher.Hash([]byte("val252"))},
	{251, hasher.Hash([]byte("val251"))},
	{250, hasher.Hash([]byte("val250"))},
	{249, hasher.Hash([]byte("val249"))},
	{248, hasher.Hash([]byte("val248"))},
	{247, hasher.Hash([]byte("val247"))},
	{15, hasher.Hash([]byte("val15"))},
}

err = smt.MultiSet(testKVData...)
if err != nil {
	t.Fatal(err)
}
```

### Changes

Notable changes:
* implement MultiSet method

### CODE REVIEW

File: options.go [hunk 1]
```diff
@@ -30,6 +31,12 @@ func DBCacheSize(size int) Option {
 	}
 }
 
+func GoRoutinePool(pool *ants.Pool) Option {
```
Comment (guagualvcha) [inline options.go:34]: Seems useless code

File: smt.go [hunk 2]
```diff
@@ -81,6 +83,13 @@ func NewBASSparseMerkleTree(hasher *Hasher, db database.TreeDB, maxDepth uint8,
 		return nil, err
 	}
 
+	if smt.goroutinePool == nil {
+		smt.goroutinePool, err = ants.NewPool(1024)
```
Comment (guagualvcha) [inline smt.go:87]: Will 1024 too big?

File: smt.go [hunk 3]
```diff
@@ -341,27 +400,105 @@ func (tree *BASSparseMerkleTree) Set(key uint64, val []byte) error {
...
 
-		tree.journal[journalKey{targetNode.depth, targetNode.path}] = targetNode
+		targetNode = parentNodes[i]
+		tree.journal.set(journalKey{targetNode.depth, targetNode.path}, targetNode)
 	}
 	tree.root = targetNode
 	return nil
 }
 
+func (tree *BASSparseMerkleTree) MultiSet(items ...Item) error {
```
Comment (guagualvcha) [inline smt.go:426]: Why the parameter is `...Item`, does it make sense if nothing to set at all? If not, let us change to slice.

**Review by guagualvcha — APPROVED** [review]

**Comment by guagualvcha** [comment]
LGTM

**Review by forcodedancing — APPROVED** [review]

File: smt.go (hunk 3 above)
Comment (j75689) [reply smt.go]: updated

File: smt.go (hunk 2 above)
Comment (j75689) [reply smt.go]: maybe "128" is enough

**Pull request merged**